### PR TITLE
nf-test side quest

### PR DESCRIPTION
**Pull request by pinin4fjords**
This is a port of the first part of @adamrtalbot 's work in https://github.com/nextflow-io/training/pull/499 for the new structure. Also corrected gitpod links to codespaces, and adjusted wording so it's more standalone.

### CODE REVIEW

**Comment by pinin4fjords** [comment]
> No cherry-pick 😢 I will never get credit.

True, I can fix that

**Comment by pinin4fjords** [comment]
> It's not a huge deal!

New structure complicated things, but added you as author on the first commit here.

File: docs/side_quests/nf-test.md [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+```console title="Output"
+> nf-test generate pipeline main.nf
+
+Load source file '/workspaces/training/side-quests/nf-test/main.nf'
+Wrote pipeline test file '/workspaces/training/side-quests/nf-test/tests/main.nf.test
+
+SUCCESS: Generated 1 test files.
+```
+
+This will create a `tests` directory with a `main.nf.test` and a `nextflow.config` file. This is our pipeline level test file. If you run `tree tests/` you should see something like this:
```
Comment (maxulysse) [inline docs/side_quests/nf-test.md:122]: does nf-test create the tests/ folder?
I think I noticed a different behavior: https://github.com/askimed/nf-test/issues/286

File: docs/side_quests/nf-test.md [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+└── nextflow.config
+```
+
+The `main.nf.test` file is our pipeline level test file. Let's open it up and take a look at the contents.
+
+```groovy title="tests/main.nf.test"
+nextflow_pipeline {
+
+    name "Test Workflow main.nf"
+    script "main.nf"
```
Comment (maxulysse) [inline docs/side_quests/nf-test.md:136]: that should not work if main.nf is one level up

File: docs/side_quests/nf-test.md [hunk 3]
```diff
@@ -0,0 +1,800 @@
...
+
+    test("Should run without failures") {
```
Comment (maxulysse) [inline docs/side_quests/nf-test.md:138]: Can we please give a better name for the test?

File: docs/side_quests/nf-test.md [hunk 4]
```diff
@@ -0,0 +1,800 @@
...
+The `nf-test` package provides an initialization command that sets up a few things in order for us to start developing tests for our project.
+
+```bash
+nf-test init
+```
+
+This should produce the following output:
+
+```bash
+🚀 nf-test 0.9.1
```
Comment (maxulysse) [inline docs/side_quests/nf-test.md:96]: can we use the same 0.9.2 version all through the document?

File: docs/side_quests/nf-test.md (hunk 1 above)
Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:122]: It does

File: docs/side_quests/nf-test.md (hunk 3 above)
Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:138]: But then we'd have to tell the users to change it, which I don't think adds anything to the message at this point

Comment (maxulysse) [reply docs/side_quests/nf-test.md:138]: we should tell users to change it.

File: docs/side_quests/nf-test.md (hunk 2 above)
Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:136]: Pretty sure it did though. I'll double check

Comment (maxulysse) [reply docs/side_quests/nf-test.md:136]: For me this and my previous comment in line 122 are linked.

Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:136]: confirmed, works just fine

File: docs/side_quests/nf-test.md (hunk 3 above)
Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:138]: What would you like them to change it to?

File: docs/side_quests/nf-test.md (hunk 2 above)
Comment (adamrtalbot) [reply docs/side_quests/nf-test.md:136]: yes it does, nf-test looks in the root of the repo unless you use a relative path.

The behaviour can change if you mess around with the `testsDir` of the `nf-test.config` file, but we want to avoid that here and just carefully introduce users to testing and nf-test.

File: docs/side_quests/nf-test.md (hunk 3 above)
Comment (adamrtalbot) [reply docs/side_quests/nf-test.md:138]: Let's run the test, then change the name of the test as an explicit learning point, highlighting how it changes in the console output. 

If I had my way I would introduce renaming tests when we have _multiple_ tests on the same process/workflow/pipeline, which is what it's intended for after all. Maybe we should introduce that in this course? We could add "test the pipeline fails if you don't specify an input", or "test zero processes run with an empty input file".

File: docs/side_quests/nf-test.md [hunk 5]
```diff
@@ -0,0 +1,832 @@
...
+
+Let's run the workflow to make sure it's working as expected.
+
+```bash
+nextflow run main.nf
+```
+
+CONGRATULATIONS! You just ran a unit test!
+
+"Wait, what? I just ran the workflow and it worked! How is that a unit test?"
```
Comment (vdauwera) [inline docs/side_quests/nf-test.md:65]: This is confusing for a couple of reasons. 

1. 'Unit test' is the only type of test not explicitly listed above, and non-programmers will typically have never heard of unit tests.
2. Running an entire workflow is not like a unit test at all; arguably the closest thing to a unit test here is a process-level test.

FWIW when I taught the original nf-test material I made the approximation that process level tests are like unit tests, and that workflow level tests are like integration tests (because they test the integration of all the processes together). I'd like to use that formally in these materials.

**Comment by vdauwera** [comment]
General comment: try to start every sentence on a new line as it makes it easier to review specific bits of text, and will make reviewing further changes easier as well.

File: docs/side_quests/nf-test.md (hunk 3 above)
Comment (vdauwera) [reply docs/side_quests/nf-test.md:138]: I agree naming the test is an important talking point, and *definitely* introduce having multiple tests on the same thing.

Would also be good to introduce different types of tests (not just identity, but also content).

File: docs/side_quests/nf-test.md [hunk 6]
```diff
@@ -0,0 +1,832 @@
...
+
+### Takeaway
+
+You know how to write tests for a Nextflow process and run them.
+
+### What's next?
+
+Learn how to run tests for everything at once!
+
+## 3. Run tests for the entire repo
```
Comment (vdauwera) [inline docs/side_quests/nf-test.md:762]: Please write repository in full rather than just repo, as that makes it more understandable for non-English speakers

File: docs/side_quests/nf-test.md (hunk 5 above)
Comment (adamrtalbot) [reply docs/side_quests/nf-test.md:65]: > and that workflow level tests are like integration tests (because they test the integration of all the processes together)

I disagree with this, for bioinformatics I would call testing on an appropriate system (e.g. an HPC, Batch, k8s), given how much of our work devops. 

This is just semantics though. We should just call it a 'test' here.

File: docs/side_quests/nf-test.md [hunk 7]
```diff
@@ -0,0 +1,832 @@
...
+
+CONGRATULATIONS! You just ran a unit test!
+
+"Wait, what? I just ran the workflow and it worked! How is that a unit test?"
+
+Good question!
+
+Let's break down what just happened.
+
+You ran the workflow with the default parameters, you confirmed it worked and you're happy with the results. This is the essence of unit testing. As you go through this course you will notice we always run the workflow at the start to confirm everything is set up correctly. This gives us the confidence to continue with the training, knowing the code started in a working state.
```
[the remark this answers is not in the document]
Comment (adamrtalbot) [reply docs/side_quests/nf-test.md:71]: Ah it's referring to how in the hello series we nearly always start with `nextflow run .`, nf-test automates this procedure.

[the remark this answers is not in the document]
Comment (adamrtalbot) [reply docs/side_quests/nf-test.md:71]: what's a better way of saying that?

File: docs/side_quests/nf-test.md (hunk 6 above)
Comment (adamrtalbot) [reply docs/side_quests/nf-test.md:762]:
```suggestion
## 3. Run tests for the entire repository
```

File: docs/side_quests/nf-test.md (hunk 5 above)
Comment (vdauwera) [reply docs/side_quests/nf-test.md:65]: Semantics is everything when we're deciding what to call things 😆 

I'm going off of what the GATK devs called integration tests, but recognize different groups use different terminologies.

So ok, let's just call this a workflow-level test or something to that effect.

File: docs/side_quests/nf-test.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (vdauwera) [reply docs/side_quests/nf-test.md:71]: Maybe something like "If you worked through the Hello Nextflow training course, you'll have noticed we always started every section by running the workflow we were using as a starting point, to confirm everything is set up correctly."

File: docs/side_quests/nf-test.md (hunk 5 above)
Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:65]: I'm with @vdauwera here actually- testing processes feels like unit testing, workflows are integration tests. 

But maybe avoiding the language altogether and just being specific on what we mean is the best way forward in any case.

Comment (pinin4fjords) [reply docs/side_quests/nf-test.md]: Just 'test' is fine here, changed accordingly

File: docs/side_quests/nf-test.md [hunk 8]
```diff
@@ -0,0 +1,1074 @@
...
+- **Transparency**: The tests show where a pipeline is failing and make it easier to track down the problem. They also function as a form of documentation, showing how to run a process or workflow.
+- **Speed**: Because the tests are automated, they can be run very quickly and repeatedly. You can iterate quickly with less fear of introducing new bugs.
+
+There are lots of different types of tests we can write:
+
+1. **Module-level tests**: For individual processes
+2. **Workflow-level tests**: For a single workflow
+3. **Pipeline-level tests**: For the pipeline as a whole
+4. **Performance tests**: For the speed and efficiency of the pipeline
+5. **Stress tests**: To identify the limits of the pipeline
```
Comment (mribeirodantas) [inline docs/side_quests/nf-test.md:24]:
```suggestion
5. **Stress tests**: Assessing the pipeline's performance under extreme conditions to determine its limits
```

File: docs/side_quests/nf-test.md [hunk 9]
```diff
@@ -0,0 +1,1074 @@
...
+```console title="Output"
+> nf-test generate pipeline main.nf
+
+Load source file '/workspaces/training/side-quests/nf-test/main.nf'
+Wrote pipeline test file '/workspaces/training/side-quests/nf-test/tests/main.nf.test
+
+SUCCESS: Generated 1 test files.
+```
+
+This will create a `tests` directory with a `main.nf.test` and a `nextflow.config` file. This is our pipeline level test file. If you run `tree tests/` you should see something like this:
```
Comment (mribeirodantas) [inline docs/side_quests/nf-test.md:130]: The `tests` directory has already been created by `nf-test init`, no? That's what is said above. I also tried it here and saw this.

```suggestion
This will create a `main.nf.test` file within the `tests` directory. This is our pipeline level test file. If you run `tree tests/` you should see something like this:
```

Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:130]: Yes, you're correct- thanks!

File: docs/side_quests/nf-test.md [hunk 10]
```diff
@@ -0,0 +1,1074 @@
...
+> nf-test generate process main.nf
+
+Load source file '/workspaces/training/side-quests/nf-test/main.nf'
+Wrote process test file '/workspaces/training/side-quests/nf-test/tests/main.sayhello.nf.test
+Wrote process test file '/workspaces/training/side-quests/nf-test/tests/main.converttoupper.nf.test
+
+SUCCESS: Generated 2 test files.
+```
+
+Let's focus for now and focus on the `sayhello` process in the `main.sayhello.nf.test` file.
```
Comment (mribeirodantas) [inline docs/side_quests/nf-test.md:538]:
```suggestion
Let's focus for now on the `sayhello` process in the `main.sayhello.nf.test` file.
```

File: docs/side_quests/nf-test.md [hunk 11]
```diff
@@ -0,0 +1,1074 @@
...
+
+  Nextflow 24.10.4 is available - Please consider updating your version to it
+
+
+
+
+FAILURE: Executed 1 tests in 4.884s (1 failed)
+```
+
+The test fails because the `sayHello` process declares 1 input channel but 0 were specified. Let's fix that by adding an input to the process. Remember from part 1, our `sayHello` process takes a single value input. We should also fix the test name to better reflect what we're testing.
```
Comment (mribeirodantas) [inline docs/side_quests/nf-test.md:616]:
```suggestion
The test fails because the `sayHello` process declares 1 input channel but 0 were specified. Let's fix that by adding an input to the process. Remember from Part 1, our `sayHello` process takes a single value input. We should also fix the test name to better reflect what we're testing.
```

We've been capitalizing `Part` in the Hello Nextflow training, so suggesting this here for consistency.

Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:616]: Thanks for highlighting. Actually, referencing part 1 isn't very helpful here, since we're on a side-quest. I probably need to re-introduce the workflow content quickly early on

File: docs/side_quests/nf-test.md [hunk 12]
```diff
@@ -0,0 +1,1074 @@
...
+  ERROR ~ No such file or directory: /workspaces/training/side-quests/nf-test/.nf-test/tests/693ba951a20fec36a5a9292ed1cc8a9f/greetings.csv
+
+   -- Check '/workspaces/training/side-quests/nf-test/.nf-test/tests/693ba951a20fec36a5a9292ed1cc8a9f/meta/nextflow.log' file for details
+  Nextflow stderr:
+
+  Nextflow 24.10.4 is available - Please consider updating your version to it
+
+
+
+
```
Comment (kenibrewer) [inline docs/side_quests/nf-test.md:235]:
```suggestion
```
Delete for clarity and to avoid distractions.

File: docs/side_quests/nf-test.md [hunk 13]
```diff
@@ -0,0 +1,1074 @@
...
+  Process `sayHello` declares 1 input channel but 0 were specified
+
+   -- Check script '/workspaces/training/side-quests/nf-test/.nf-test-1eaad118145a1fd798cb07e7dd75d087.nf' at line: 38 or see '/workspaces/training/side-quests/nf-test/.nf-test/tests/1eaad118145a1fd798cb07e7dd75d087/meta/nextflow.log' file for more details
+  Nextflow stderr:
+
+  Nextflow 24.10.4 is available - Please consider updating your version to it
+
+
+
+
```
Comment (kenibrewer) [inline docs/side_quests/nf-test.md:612]:
```suggestion
```
Delete for clarity and to avoid distractions.

**Review by kenibrewer — APPROVED** [review]
This is really good material and is ready in my view once the minor typos that @mribeirodantas and I flagged are addressed.

File: docs/side_quests/nf-test.md (hunk 12 above)
Comment (pinin4fjords) [reply docs/side_quests/nf-test.md:295]: Sorry, yes, didn't mean to include

**Review by adamrtalbot — APPROVED** [review]
🎉

File: docs/side_quests/nf-test.md [hunk 14]
```diff
@@ -0,0 +1,1124 @@
...
+You know how to write tests for a Nextflow process and run them.
+
+### What's next?
+
+Learn how to run tests for everything at once!
+
+## 3. Run tests for the entire repository
+
+Running nf-test on each component is fine, but laborious and error prone. Can't we just test a test on everything at once?
```
Comment (mribeirodantas) [inline docs/side_quests/nf-test.md:1047]:
```suggestion
Running nf-test on each component is fine, but laborious and error prone. Can't we just test everything at once?
```
What do you think?

**Review by mribeirodantas — DISMISSED** [review]
Amazing work, @adamrtalbot and @pinin4fjords ! I'm happy with this PR getting merged, but I would like to share the feeling I still remember having when I started playing with nf-test.

The person who does this material for the first time will finish with the feeling of "Ok, I know what are Nextflow tests now, but I still can't do one 😓".

I think we should spend more time helping people provide real inputs to tests, or using some operators to do something with channel elements. It doesn't have to be complex, just more examples so that the person has more material to work with when they want to write their tests. This could be added later, but I think it's important to add at some point.

PS: Some comments on the setup block too, as @vdauwera mentioned.

**Review by mribeirodantas — DISMISSED** [review]

**Comment by mribeirodantas** [comment]
LGTM?

**Comment by mribeirodantas** [comment]
I added line numbers / line highlights to the code block comparisons. More details in the commit message. Feel free to revert it if you think it's not helpful.

**Review by mribeirodantas — APPROVED** [review]

**Comment by pinin4fjords** [comment]
> I added line numbers / line highlights to the code block comparisons. More details in the commit message. Feel free to revert it if you think it's not helpful.

I like it!

**Comment by pinin4fjords** [comment]
> PS: Some comments on the setup block too, as @vdauwera mentioned.

I'm not 100% sold on including setup at this stage. It's not really needed until you get properly into the weeds with nf-test, I think it might muddy the waters to introduce too early.

**Comment by vdauwera** [comment]
> > PS: Some comments on the setup block too, as @vdauwera mentioned.
> 
> I'm not 100% sold on including setup at this stage. It's not really needed until you get properly into the weeds with nf-test, I think it might muddy the waters to introduce too early.

Yeah on further thought I'd be ok leaving it for an 'intermediate' level add-on to this side quest.

**Comment by mribeirodantas** [comment]
> Yeah on further thought I'd be ok leaving it for an 'intermediate' level add-on to this side quest.

I agree, but we definitely need the next one! My feeling is that this first piece only gives an idea of nf-test, and is not really applicable to what people may need, even in simple pipelines. And that's fine, imho, as long as we have next step 😆

**Review by mribeirodantas — APPROVED** [review]
